### PR TITLE
Add teaching elements sidebar for metadata entry

### PR DESCRIPTION
**Pull request by mbareta**
This will add `meta` field to `te.model` and a sidebar for metadata input.
The `meta` is configurable by editing `tesMeta` field in schema config. [Example here](https://github.com/ExtensionEngine/tailor/pull/129/files#diff-44ebac34859059a430ce4631f2bd6ff7R86)
TES meta supports the same options like Activity meta.

### CODE REVIEW

File: client/components/course/Sidebar/Body.vue [hunk 1]
```diff
@@ -54,9 +50,9 @@ export default {
     };
   },
   computed: {
-    ...mapGetters(['activity'], 'course'),
+    ...mapGetters(['activity', 'focusedElementConfig', 'focusedElementMetadata'], 'course'),
     config() {
-      return getLevel(this.activity.type) || {};
+      return this.focusedElementConfig();
```
Comment (vladimyr) [inline client/components/course/Sidebar/Body.vue]:
```js
return this.getConfig(this.activity);
```

File: client/components/course/Sidebar/Body.vue [hunk 2]
```diff
@@ -54,9 +50,9 @@ export default {
     };
   },
   computed: {
-    ...mapGetters(['activity'], 'course'),
+    ...mapGetters(['activity', 'focusedElementConfig', 'focusedElementMetadata'], 'course'),
```
Comment (vladimyr) [inline client/components/course/Sidebar/Body.vue]:
```js
...mapGetters(['activity', 'getConfig', 'getMetadata'], 'course')
```

File: client/components/course/Sidebar/Body.vue [hunk 3]
```diff
@@ -65,18 +61,13 @@ export default {
         : 'Not published';
     },
     metadata() {
-      if (!get(this.config, 'meta')) return [];
-      return map(this.config.meta, it => {
-        let value = get(this.activity, `data.${it.key}`);
-        return { ...it, value };
-      });
+      return this.focusedElementMetadata();
```
Comment (vladimyr) [inline client/components/course/Sidebar/Body.vue]:
```js
return this.getMetadata(this.activity);
```

File: client/components/editor/index.vue [hunk 4]
```diff
@@ -37,13 +54,17 @@ export default {
...
+      showSidebar: false,
       mousedownCaptured: false
     };
   },
   computed: {
     ...mapGetters(['activities']),
-    ...mapGetters(['focusedElement', 'activity', 'contentContainers'], 'editor'),
-    ...mapGetters(['course'], 'course'),
+    ...mapGetters(['activity', 'contentContainers', 'focusedElement'], 'editor'),
+    ...mapGetters(['course', 'focusedElementMetadata'], 'course'),
```
Comment (vladimyr) [inline client/components/editor/index.vue]:
```js
...mapGetters(['course', 'getMetadata'], 'course'),
```

File: client/components/editor/index.vue [hunk 5]
```diff
@@ -37,13 +54,17 @@ export default {
...
+    metadata() {
+      return this.focusedElementMetadata(true);
```
Comment (vladimyr) [inline client/components/editor/index.vue]:
```js
return this.getMetadata(this.focusedElement);
```

File: client/components/editor/index.vue [hunk 6]
```diff
@@ -85,6 +106,11 @@ export default {
       }
     }
   },
+  watch: {
+    focusedElement(next, previous) {
+      if (next._cid !== previous._cid) this.showSidebar = false;
```
Comment (vladimyr) [inline client/components/editor/index.vue]: Next sounds weird. It is current actually :bowtie:

File: client/store/modules/course.js [hunk 7]
```diff
@@ -67,6 +69,32 @@ getter(function revisions() {
     .sort((rev1, rev2) => rev2.createdAt - rev1.createdAt);
 });
 
+getter(function getConfig() {
+  return element => {
+    if (!element.type) return {};
+    // NOTE: teaching elements always have activityId foreign key and that is
+    // how we can tell if an element is tes or activity
+    const isTes = !!element.activityId;
```
Comment (vladimyr) [inline client/store/modules/course.js]: Just promote it to top level helper by converting to lambda :bowtie:

File: client/store/modules/course.js [hunk 8]
```diff
@@ -67,6 +69,32 @@ getter(function revisions() {
...
+    if (isTes) {
+      const course = this.rootGetters['course/course'];
+      return getTesMeta(course.schema, element.type) || {};
+    }
+    return getLevel(element.type) || {};
```
Comment (vladimyr) [inline client/store/modules/course.js]: It would be nice if possible to add optional last param for default value :bulb:

**Comment by kjuej** [comment]
@mbareta When editing sidebar meta before saving the original content changes, the original content changes are reset.

File: client/store/modules/course.js [hunk 9]
```diff
@@ -67,6 +70,30 @@ getter(function revisions() {
     .sort((rev1, rev2) => rev2.createdAt - rev1.createdAt);
 });
 
+getter(function getConfig() {
+  return element => {
+    if (!element.type) return {};
+    // NOTE: teaching elements always have activityId foreign key and that is
+    // how we can tell if an element is tes or activity
```
Comment (vladimyr) [inline client/store/modules/course.js]: Move this comment above `isTes` declaration ⬆️

File: client/store/modules/course.js (hunk 8 above)
Comment (vladimyr) [reply client/store/modules/course.js]: Hmm, idea was to do following:
```js
function getTesMeta(schemaId, type, defaultValue) {
  return find(getSchema(schemaId).tesMeta, { type }) || defaultValue;
}

getTesMeta(course.schema, element.type, {});
// and modify getLevel in same fashion:
getLevel(element.type, {});
```

**Comment by mbareta** [comment]
@kjuej ready for another pass. Everything should work now.

**Comment by kjuej** [comment]
@mbareta Everything seems in order now.

**Comment by vladimyr** [comment]
@mbareta I reverted 3f671a97 because there where still occurrences in codebase  with unsafe `getLevel` invocations: https://github.com/ExtensionEngine/tailor/blob/a3ed7080875d6cc5b89c73f45763550106d8bd6e/config/shared/activities.js#L43-L44
So I decided, out of precaution, to provoke luck another time 🎲 🤠

**Pull request merged**